crates/sui-framework-build/src/compiled_package.rs

```diff
@@ -474,3 +491,65 @@ impl PackageHooks for SuiPackageHooks {
         Ok(())
     }
 }
+
+enum AddressError {
+    NoPublishedAddress(Symbol),
+    InvalidPublishedAddress(Symbol, String),
+}
+
+/// Ensure dependencies are published (contain `published-at` address in manifest).
+pub fn ensure_published_dependencies(resolution_graph: &ResolvedGraph) -> Result<(), SuiError> {
+    let mut address_errors = vec![];
+    for name in resolution_graph.graph.package_table.keys() {
+        if let Some(package) = &resolution_graph.package_table.get(name) {
+            let value = package
+                .source_package
+                .package
+                .custom_properties
+                .get(&Symbol::from(PUBLISHED_AT_MANIFEST_FIELD));
+
+            if value.is_none() {
+                address_errors.push(AddressError::NoPublishedAddress(*name));
+                continue;
+            }
+
+            if value
+                .and_then(|v| ObjectID::from_str(v.as_str()).ok())
+                .is_none()
+            {
+                address_errors.push(AddressError::InvalidPublishedAddress(
+                    *name,
+                    value.unwrap().to_string(),
+                ));
+                continue;
+            }
+        }
+    }
+    if !address_errors.is_empty() {
+        let mut error_msgs = vec![];
+        for kind in address_errors {
+            match kind {
+                AddressError::NoPublishedAddress(name) => error_msgs.push(format!(
+                    "Package dependency \"{name}\" does not \
+		     specify a published address \
+		     (the Move.toml manifest for \"{name}\" does \
+		     not contain a published-at field)."
+                )),
+                AddressError::InvalidPublishedAddress(name, value) => error_msgs.push(format!(
+                    "Package dependency \"{name}\" does not \
+		     specify a valid published address: \
+		     could not parse value \"{value}\" for published-at field.",
+                )),
+            }
+        }
+        error_msgs.push(
+            "If this is intentional, use the \
+	     --with-unpublished-dependencies flag to skip this check."
```

Comment:
Worth clarifying here that `--with-unpublished-dependencies` not only skips the check, but will also publish those dependencies as part of this package.

Reply:
very good, thoughts on updated wording?

Reply:
New wording looks good, made some suggestions for minor tweaks, but the sentiment is there!